pkg/skaffold/kubernetes/portforward/kubectl_forwarder.go

```diff
@@ -189,3 +210,90 @@ func (*KubectlForwarder) monitorErrorLogs(ctx context.Context, logs io.Reader, c
 		}
 	}
 }
+
+// findNewestPodForService queries the cluster to find a pod that fulfills the given service, giving
+// preference to pods that were most recently created.  This is in contrast to the selection algorithm
+// used by kubectl (see https://github.com/GoogleContainerTools/skaffold/issues/4522 for details).
+func findNewestPodForService(ctx context.Context, ns, serviceName string, servicePort int) (string, int, error) {
+	client, err := kubernetes.Client()
+	if err != nil {
+		return "", -1, fmt.Errorf("getting Kubernetes client: %w", err)
+	}
+	svc, err := client.CoreV1().Services(ns).Get(serviceName, metav1.GetOptions{})
+	if err != nil {
+		return "", -1, fmt.Errorf("getting service: %w", err)
+	}
+	svcPort, err := findServicePort(*svc, servicePort)
+	if err != nil {
+		return "", -1, err
+	}
+
+	set := labels.Set(svc.Spec.Selector)
+	listOptions := metav1.ListOptions{LabelSelector: set.AsSelector().String()}
+	pods, err := client.CoreV1().Pods(ns).List(listOptions)
+	if err != nil {
+		return "", -1, fmt.Errorf("listing pods: %w", err)
+	}
+	sort.Slice(pods.Items, newestPodsFirst(pods.Items))
+
+	if logrus.IsLevelEnabled((logrus.TraceLevel)) {
+		var names []string
+		for _, p := range pods.Items {
+			names = append(names, fmt.Sprintf("(pod:%q phase:%v created:%v)", p.Name, p.Status.Phase, p.CreationTimestamp))
+		}
+		logrus.Tracef("service %s/%d maps to %d pods: %v", serviceName, servicePort, len(pods.Items), names)
+	}
+
+	for _, p := range pods.Items {
+		if targetPort := findTargetPort(svcPort, p); targetPort > 0 {
+			logrus.Debugf("Forwarding service %s/%d to pod %s/%d", serviceName, servicePort, p.Name, targetPort)
+			return p.Name, targetPort, nil
+		}
+	}
+
+	return "", -1, fmt.Errorf("no pods match service %s/%d", serviceName, servicePort)
+}
+
+// newestPodsFirst sorts pods by their running state and then by creation time.
+func newestPodsFirst(pods []corev1.Pod) func(int, int) bool {
+	// prefer Running then Pending
+	phases := map[corev1.PodPhase]int{
+		corev1.PodRunning:   0,
+		corev1.PodPending:   1,
```

Comment:
> `corev1.PodPending:   1,`
is it possible that there's some race where a new pod is still pending while the old one returns running?

Reply:
That situation is no different than what happens with `kubectl port-forward service/xxx`.

Perhaps I should be further refining the ListOptions to filter by pod-phase.

Reply:
Along the same lines, would it make sense to always select the most recent pod, not matter its state. It feels more predictable. And if this pod fails to start a bit later, bad things will happen anyways.

Reply:
I'm fine with treating Pending and Running as equivalent — the others mean the containers are terminated and so we shouldn't send them there.  Let me look at the ListOptions.

Reply:
With status-checks enabled, we generally wait for pending pods to become running, but without status-checks we'll be more likely to hit pending pods.  

ListOptions field-selectors only effective since Kubernetes 1.16 so a no go.  I'll filter for Pending and Running pods and treat them as equivalent.